docs/architecture/server/index.md

```diff
@@ -0,0 +1,119 @@
+always need their own class.
+
+The class, interface and public method should be named after the action. For example:
+
+```csharp
+namespace Bit.Core.OrganizationFeatures.OrganizationApiKeys;
+
+public class RotateOrganizationApiKeyCommand : IRotateOrganizationApiKeyCommand
+{
+  public async Task<OrganizationApiKey> RotateApiKeyAsync(OrganizationApiKey organizationApiKey)
+  {
+    ...
+  }
+}
+```
+
+The query/command should only expose public methods that run the complete action. It should not have
+public helper methods.
+
+The directory structure and namespaces should be organized by feature. Interfaces should be stored
+in a separate sub-folder. For example:
+
+```text
+  Core/
+    └── OrganizationFeatures/
+        └── OrganizationApiKeys/
+            ├── Interfaces/
+            │   └── IRotateOrganizationApiKeyCommand.cs
+            └── RotateOrganizationApiKeyCommand.cs
+```
+
+### Maintaining the command/query distinction
+
+By separating read and write operations, CQRS encourages us to maintain loose coupling between
+classes. There are two golden rules to follow when using CQRS in our codebase:
+
+- **Commands should never read and queries should never write**
+- **Commands and queries should never call eachother**
+
+Both of these lead to tight coupling between classes, reduce opportunities for code re-use, and
+conflate the command/query distinction.
+
+You can generally avoid these problems by:
+
+- writing your commands so that they receive all the data they need in their arguments, rather than
+  fetching the data themselves
+- calling queries and commands sequentially (one after the other), passing the results along the
+  call chain
+
+For example, if we need to update an API key for an organization, it might be tempting to have an
+`UpdateApiKeyCommand` which fetches the current API key and then updates it. However, we can break
+this down into two separate queries/commands, which are called separately:
+
+```csharp
+var currentApiKey = await _getOrganizationApiKeyQuery.GetOrganizationApiKeyAsync(orgId);
+await _rotateOrganizationApiKeyCommand.RotateApiKeyAsync(currentApiKey);
+```
+
+This has unit testing benefits as well - instead of having lengthy "arrange" phases where you mock
+query results, you can simply supply different argument values using the `Autodata` attribute.
+
+### Avoid Primitive Obsession
+
+Where practical, your commands and queries should take and return whole objects (e.g. `User`) rather
+than individual properties (e.g. `userId`).
+[Read more about Primitive Obsession](https://refactoring.guru/smells/primitive-obsession).
+
+### Avoid excessive optional parameters
+
+Lots of optional parameters can quickly become difficult to work with. Instead, consider using
+method overloading to provide different entry points into your command or query.
+
+### Transition plan
+
+We are gradually transitioning to a CQRS pattern over time. If you're making changes that use or
```

Comment:
ℹ️ Similar comment -- best to speak to the intention of what "is" and move and of this transitional content to the ADR.

Reply:
My thinking was that most people will not read the ADR (which is basically a historical document), and we need as many devs as possible to contribute to migrating existing logic. So it's useful here for visibility. I've gone ahead and done it anyway though, in case this doesn't convince you.